src/fetchers/onchain.js

```diff
@@ -0,0 +1,199 @@
+  try {
+    const response = await fetch(RPC_URL, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(createBatch(encodedTokens)),
+      timeout: FETCH_TIMEOUT,
+      // TODO: add proxy support to avoid rate limiting
+      // agent:
+    });
+    const json = await response.json();
+    return [json, null];
+  } catch (e) {
+    return [null, e.message];
+  }
+};
+
+const getTokenMetadataFromURI = async (uri) => {
+  try {
+    if (uri.includes("ipfs://")) {
+      uri = uri.replace("ipfs://", "https://ipfs.io/ipfs/");
+    }
+    if (!uri.includes("http")) {
+      return [null, "Invalid URL"];
+    }
+
+    const response = await fetch(uri, {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      timeout: FETCH_TIMEOUT,
+      // TODO: add proxy support to avoid rate limiting
+      // agent:
+    });
+
+    if (!response.ok) {
+      return [null, response.status];
+    }
+
+    const json = await response.json();
+    return [json, null];
+  } catch (e) {
+    return [null, e.message];
+  }
+};
+
+export const fetchTokens = async (chainId, tokens, standard = "ERC721") => {
+  // TODO: Add support for other chains via RPC_URL
+  if (tokens.length === 0) return [];
+  if (!Array.isArray(tokens)) tokens = [tokens];
+  //   if (tokens.length > 20) throw new Error("Too many tokenIds (max 20)");
+  if (chainId !== 1) throw new Error("Only mainnet is supported");
+
+  // We need to have some type of hash map to map the tokenid + contract to the tokenURI
+  const idToToken = {};
+  tokens.forEach((token) => {
+    const randomInt = Math.floor(Math.random() * 100000);
+    idToToken[randomInt] = token;
+    token.requestId = randomInt;
+  });
+
+  const RPC_URL = process.env.MAINNET_RPC_URL;
+  const encodeTokenFunction = standard === "ERC721" ? encodeTokenERC721 : encodeTokenERC1155;
+  const encodedTokens = tokens.map(encodeTokenFunction);
+  const [batch, error] = await sendBatch(encodedTokens, RPC_URL);
+  if (error) {
+    return tokens.map((token) => {
+      return {
+        contract: token.contract,
+        token_id: token.tokenId,
+        error: "Unable to fetch tokenURI from contract",
+      };
+    });
+  }
+
+  const resolvedMetadata = await Promise.all(
+    batch.map(async (token) => {
+      try {
+        const uri = Web3Local.eth.abi.decodeParameter("string", token.result);
+        const [metadata, error] = await getTokenMetadataFromURI(uri);
+        if (error) {
+          return {
+            contract: idToToken[token.id].contract,
+            token_id: idToToken[token.id].tokenId,
+            error: "Unable to fetch metadata from URI",
+          };
+        }
+
+        return {
+          ...metadata,
+          contract: idToToken[token.id].contract,
+          token_id: idToToken[token.id].tokenId,
+        };
+      } catch (e) {
+        return {
```

Comment:
Can we detect a rate limit error and respond appropriately (see [here](https://github.com/reservoirprotocol/metadata/blob/cdbbed9b76663a86ebd57ee38048b95687da0315/src/fetchers/opensea.js#L274)). This will make sure any tokens that are pending refresh wont get discarded and be put back on the queue on the indexer side.